docs/adr/ADR-004-centralized-rate-limiting-via-hooks.md

```diff
@@ -0,0 +1,276 @@
+# ADR-004: Centralized Rate Limiting via API Wrapper Hooks
+
+**Status:** Accepted
+**Date:** 2025-11-14
+**Authors:** cassing
+**Supersedes:** N/A
+**Superseded by:** N/A
+
+## Context
+
+External APIs (Slack, GitHub, GitLab, AWS) have rate limits:
+
+- **Slack**: Tier-based rate limits (20-100+ requests/minute)
+- **GitHub**: 5000 requests/hour for authenticated, lower for unauthenticated
+- **GitLab**: 300 requests/minute per user
+- **AWS**: Service-specific limits (varies by service/account)
+
+### Problem
+
+Where should rate limiting be implemented?
+
+**Multiple consumers exist:**
+
+1. **qontract-api service layer**: Reconciliation endpoints
+2. **Direct integrations**: Existing `reconcile/slack_usergroups.py`
+3. **Manual scripts**: Admin tools, debugging scripts
+4. **Future consumers**: Other services using SlackApi
+
+**Challenges:**
+
+- Rate limits apply to the workspace/account, not per consumer
+- Multiple processes might use same API (distributed rate limiting needed)
+- Different tiers have different limits (configuration needed)
+- Code duplication if each consumer implements own rate limiting
```

Comment:
The root problem is when remote service rate limit our requests, we don't have a feedback loop to slow down sending requests to remote, lead to more errors on our side and slower recovery from rate limiting.

The solution should be react to remote rate limiting by adjusting our request rate, longer backoff on retry.

Reply:
This could be easily implemented via an `after_request` hook. Let's postpone that for now. We need a "feature freeze" for the POC; otherwise, it won't go live ever.